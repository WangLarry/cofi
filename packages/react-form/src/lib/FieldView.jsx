/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import * as Styled from './Styled';

const propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  description: PropTypes.string,
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ render: PropTypes.func.isRequired }),
  ]),
  value: PropTypes.any,
  state: PropTypes.object,
  excluded: PropTypes.bool,
  disabled: PropTypes.bool,
  dirty: PropTypes.bool,
  required: PropTypes.bool,
  empty: PropTypes.bool,
  invalid: PropTypes.bool,
  errors: PropTypes.array,
  styleOverrides: PropTypes.string,
  onValueChange: PropTypes.func,
  onStateChange: PropTypes.func,
};

const FieldView = ({
  id = undefined,
  label = undefined,
  description = undefined,
  value = undefined,
  excluded = false,
  disabled = false,
  dirty = false,
  required = false,
  empty = false,
  invalid = false,
  errors = [],
  state = {},
  component = undefined,
  styleOverrides = undefined,
  onValueChange = noop,
  onStateChange = noop,
  ...customProps
}) => {
  if (excluded) return null;

  const Component = component;

  return (
    <Styled.Field 
      id={id}
      aria-label="Field" 
      data-disabled={disabled} 
      data-dirty={dirty} 
      data-required={required}
      data-empty={empty} 
      data-invalid={invalid}
      styleOverrides={styleOverrides}>
      {
        label && <div aria-label="Header">
          <div aria-label="Label">{label}</div>
          { required && (<div aria-label="Required">*</div>) }
          { description && (<i aria-label="Description" title={description}>?</i>) }
        </div>
      }
      {
        Component && <div aria-label="Component">
          <Component
            {...customProps}
            value={value}
            state={state}
            disabled={disabled}
            dirty={dirty}
            required={required}
            empty={empty}
            invalid={invalid} 
            onValueChange={onValueChange}
            onStateChange={onStateChange} />
        </div>
      }
      {
        errors.length > 0 && <div aria-label="Error">{errors[0].message}</div>
      }
    </Styled.Field>
  );
};

FieldView.prototype = propTypes;

export default FieldView;
