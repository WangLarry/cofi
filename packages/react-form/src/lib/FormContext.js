/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';

const FormContext = React.createContext();
export default FormContext;
