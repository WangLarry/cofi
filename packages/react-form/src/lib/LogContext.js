/**
  * Copyright 2020, Gal Havivi
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';

const LogContext = React.createContext();
export default LogContext;
