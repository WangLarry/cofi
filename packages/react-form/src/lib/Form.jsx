/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React, { useState, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import JForm from '@cofi/form';
import FormContext from './FormContext';

const Form = ({ model, resources, settings, data, context, children }) => {
  const [form, setForm] = useState();
  const [privates] = useState({ models: [] });
  const parent = useContext(FormContext);

  const safeDestroy = async () => {
    // if last init was a success (not failed on definition error)
    if (privates.form && !privates.form.destroyed && privates.lastInitSuccess) {
      await privates.form.destroy();
    }
  };

  // on init and on model prop changes - init
  // queue the changed models, for cases models was changed few times in a row without finish its first init
  useEffect(() => {
    const updatePublicForm = (form) => !privates.destroyed && setForm(form);

    const init = async () => {
      privates.models.push(model);

      if (!privates.processing) {
        privates.processing = true;

        while (privates.models.length) {
          await safeDestroy();
          privates.form = new JForm();
          const currModel = privates.models[0];
          const contextOverride = context ? { context: context } : undefined;
          const formModel = Object.assign({}, currModel, contextOverride);
          privates.lastInitSuccess = await privates.form.init(formModel, resources, settings, updatePublicForm);
          privates.models.shift();
        }
      
        privates.processing = false;
      }
    };

    init();
  }, [model]); // eslint-disable-line

  // on un-mount - mark destroyed, to disable upcoming ui dispatches
  useEffect(() => {
    return () => {
      safeDestroy();
      privates.destroyed = true; 
    };
  }, []); // eslint-disable-line

  // when data prop changes - call changeData
  useEffect(() => { 
    privates.firstTimeData && privates.form.changeData(data);
    privates.firstTimeData = true;
  }, [data]); // eslint-disable-line

  // when context prop changes - call changeData
  useEffect(() => {
    privates.firstTimeContext && privates.form.changeContext(context);
    privates.firstTimeContext = true;
  }, [context]); // eslint-disable-line

  if (!form) return (null);

  const contextValue = {
    parent,
    model: form.model,
    resources: form.resources,
    actions: {
      changeValue: privates.form.changeValue.bind(privates.form),
      changeData: privates.form.changeData.bind(privates.form),
      changeContext: privates.form.changeContext.bind(privates.form),
      changeState: privates.form.changeState.bind(privates.form),
      changeUi: privates.form.changeUi.bind(privates.form),
      submit: privates.form.submit.bind(privates.form),
      reset: privates.form.reset.bind(privates.form),
    },
  };

  return (<FormContext.Provider value={contextValue}><>{children}</></FormContext.Provider>);
};

Form.prototype = {
  model: PropTypes.object.isRequired,
  resources: PropTypes.object,
  settings: PropTypes.object,
  data: PropTypes.object,
  context: PropTypes.object,
};

export default Form;
