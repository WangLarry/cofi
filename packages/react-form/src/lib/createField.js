/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React, { useState, useContext } from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';
import { mapFieldToProps } from '@cofi/form';
import FormContext from './FormContext';

export default function createField(FieldView) {

  const Field = ({ id, ...customProps }) => {
    const [cachedProps, setCachedProps] = useState();
    const [cachedFieldView, setCachedFieldView] = useState((null));
    const { model, resources, actions } = useContext(FormContext);
    
    const props = mapFieldToProps(id, model, resources);
  
    // optimization prevent un-necessary field renders
    if (!isEqual(props, cachedProps)) {
      setCachedProps(props);
      setCachedFieldView(<FieldView
        {...customProps}
        {...props}
        onValueChange={value => actions.changeValue(id, value)}
        onStateChange={state => actions.changeState(id, state)}
      />);
    }

    return cachedFieldView;
  };
  
  Field.propTypes = {
    id: PropTypes.string.isRequired,
  };

  return Field;
}
