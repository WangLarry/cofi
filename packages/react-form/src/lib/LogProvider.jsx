/**
  * Copyright 2020, Gal Havivi
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React, { useState, useEffect, useCallback } from 'react';
import cloneDeep from 'lodash/cloneDeep';
import { setLogger, Steps } from '@cofi/form';
import LogContext from './LogContext';

const defaultSettings = { 
  debugMaxLength: 50, 
  errorMaxLength: 10, 
  warnMaxLength: 10,
  formIds: [],
};

const initialHistory = { debug: [], warn: [], error: [] };

const addRecord = (history, funcName, record, maxLength) => {
  history[funcName].push(record);

  if (history[funcName].length > maxLength) {
    history[funcName].splice(0, 1);
  }
};

const addRecordAndUpdateLog = (history, setHistory, funcName, record, maxLength, shouldSet = () => true) => {
  addRecord(history, funcName, record, maxLength);

  // optimization - react render for historys component only on specific condition
  if (shouldSet()) {
    setTimeout(() => { // optimization - make page loads faster, since debug message called during init action
      setHistory({ 
        debug: history.debug, 
        warn: history.warn,
        error: history.error,
      });
    });
  }
};

// step exists only for debug records
const getRecordFormId = ({ form, step = {} }) => form && form.model ? form.model.id : step.formId;

const shouldAdd = (record, settings) => !settings.formIds.length ||
  settings.formIds.includes(getRecordFormId(record));

const customLogger = (history, setHistory, settings) => ({
  error: record => shouldAdd(record, settings)
    && addRecordAndUpdateLog(history, setHistory, 'error', record, settings.errorMaxLength),
  warn: record => shouldAdd(record, settings)
    && addRecordAndUpdateLog(history, setHistory, 'warn', record, settings.warnMaxLength),
  debug: record => shouldAdd(record, settings)
    && addRecordAndUpdateLog(history, setHistory, 'debug', record, settings.debugMaxLength,
      () => record.step.type === Steps.END_PROCESSING),
});

const generateId = () => Math.random().toString(16).slice(2);

export default (props) => {
  const [history, setHistory] = useState(cloneDeep(initialHistory));
  const [settings, setSettings] = useState();

  useEffect(() => {
    const _settings = { ...defaultSettings, ...props.settings };
    const name = generateId(); // unique between multiple log provider
    setLogger(name, customLogger(history, setHistory, _settings));
    setSettings(_settings);

    // cleanup
    return () => setLogger(name);
  }, [history, setHistory, props.settings]);

  const clear = useCallback(() => setHistory(cloneDeep(initialHistory)), [setHistory]);

  if (!settings) return (null);

  const context = {
    settings,
    history,
    clear,
  };

  return (<LogContext.Provider value={context}>{ props.children }</LogContext.Provider>);
};
