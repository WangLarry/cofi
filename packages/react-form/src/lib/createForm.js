/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import Form from './Form';

export default function createForm({ model, resources, settings }) {
  return Inner => props => {
    const { context, data, ...customProps } = props;
    return (<Form model={model} resources={resources} settings={settings} context={context} data={data}>
      <Inner { ...customProps } />
    </Form>);
  };
}
