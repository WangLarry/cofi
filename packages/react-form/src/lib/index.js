/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import FormContext from './FormContext';
import FieldView from './FieldView';
import Field from './Field';
import Form from './Form';
import LogContext from './LogContext';
import LogProvider from './LogProvider';
import createForm from './createForm';
import createField from './createField';
import withForm from './withForm';
import withLog from './withLog';

export {
  FormContext,
  Form,
  Field,
  FieldView,
  LogContext,
  LogProvider,
  createForm,
  createField,
  withForm,
  withLog,
};
