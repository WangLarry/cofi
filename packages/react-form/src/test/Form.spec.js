/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { create, act } from 'react-test-renderer';
import * as JForm from '@cofi/form';
import { Form } from '../lib';

describe('Form', () => {
  const children = <div>Mock Child</div>;
  let props;
  let mocks;
  let mocksInvalid;
  let root;

  beforeEach(() => {
    props = { 
      model: { id: 'mockId' }, 
      settings: { maxSomething: 23 },
      resources: { validators: {} },
    };

    mocks = {
      init: jest.fn(() => true), // init finish ok
      changeValue: jest.fn(),
      changeState: jest.fn(),
      changeUi: jest.fn(),
      changeData: jest.fn(),
      changeContext: jest.fn(),
      destroy: jest.fn(),
    };

    mocksInvalid = {
      init: jest.fn(() => undefined), // init finish not ok
    };

    JForm.default.prototype.init = mocks.init;
    JForm.default.prototype.changeValue = mocks.changeValue;
    JForm.default.prototype.changeState = mocks.changeState;
    JForm.default.prototype.changeUi = mocks.changeUi;
    JForm.default.prototype.changeData = mocks.changeData;
    JForm.default.prototype.changeContext = mocks.changeContext;
    JForm.default.prototype.destroy = mocks.destroy;
  });

  const init = async () => {
    // creates the form, and triggers first useEffect
    await act(async () => { root = await create(<Form {...props}>{children}</Form>); });

    // verify form render
    expect(root.toJSON()).toEqual(null);

    // verify calls
    expect(mocks.init).toHaveBeenCalledWith(props.model, props.resources, props.settings, expect.any(Function));
    expect(mocks.destroy).not.toHaveBeenCalled();
    expect(mocks.changeData).not.toHaveBeenCalled();
    expect(mocks.changeContext).not.toHaveBeenCalled();
  };
  
  it('init - triggers init action and update public ui form', async () => { 
    // init and verify
    await init();

    // mock update public form when init finish
    const updatePublicForm = mocks.init.mock.calls[0][3];
    const form = { model: props.model, settings: props.settings, resources: props.resources };
    act(() => { updatePublicForm(form); });
    
    // verify form render
    expect(root.toJSON()).toMatchSnapshot();

    // verify all correct data passed from the context value
    // TODO
  });

  it('init - with app context prop', async () => { 
    props.context = { a: 1 };
    const expectedModel = Object.assign({}, props.model, { context: props.context });

    // creates the form, and triggers first useEffect
    await act(async () => { root = await create(<Form {...props}>{children}</Form>); });

    // verify form render
    expect(root.toJSON()).toEqual(null);

    // verify calls
    expect(mocks.init).toHaveBeenCalledWith(expectedModel, props.resources, props.settings, expect.any(Function));
    expect(mocks.destroy).not.toHaveBeenCalled();
    expect(mocks.changeData).not.toHaveBeenCalled();
    expect(mocks.changeContext).not.toHaveBeenCalled();
  });

  it('change model prop - triggers destroy if needed and init actions', async () => {
    // init and verify
    await init();

    // change model prop to invalid model - triggers destroy & init actions
    JForm.default.prototype.init = mocksInvalid.init; // init finish not ok
    props.model = {};
    await act(async () => { await root.update(<Form {...props}>{children}</Form>); });
    expect(mocks.destroy.mock.calls).toHaveLength(1);
    expect(mocksInvalid.init.mock.calls).toHaveLength(1);
    expect(mocksInvalid.init.mock.calls[0]).toEqual([props.model, props.resources, props.settings, expect.any(Function)]);
    expect(mocks.init.mock.calls).toHaveLength(1); // from last time
    expect(mocks.changeContext.mock.calls).toHaveLength(0); // from last time
    expect(mocks.changeData.mock.calls).toHaveLength(0); // from last time

    // change model prop to a valid model - triggers only init action (skip destroy since last init was not a success)
    JForm.default.prototype.init = mocks.init; // init finish ok
    props.model = { id: 'test', fields: { a: { path: 'a' } } };
    await act(async () => { await root.update(<Form {...props}>{children}</Form>); });
    expect(mocks.destroy.mock.calls).toHaveLength(1); // from last time
    expect(mocksInvalid.init.mock.calls).toHaveLength(1); // from last time
    expect(mocks.init.mock.calls).toHaveLength(2);
    expect(mocks.init.mock.calls[1]).toEqual([props.model, props.resources, props.settings, expect.any(Function)]);
    expect(mocks.changeContext.mock.calls).toHaveLength(0); // from last time
    expect(mocks.changeData.mock.calls).toHaveLength(0); // from last time

    // change model prop to valid model again - triggers destroy & init actions
    props.model = { id: 'test2', fields: { b: { path: 'b' } } };
    await act(async () => { await root.update(<Form {...props}>{children}</Form>); });
    expect(mocks.destroy.mock.calls).toHaveLength(2);
    expect(mocksInvalid.init.mock.calls).toHaveLength(1); // from last time
    expect(mocks.init.mock.calls).toHaveLength(3);
    expect(mocks.init.mock.calls[2]).toEqual([props.model, props.resources, props.settings, expect.any(Function)]);
    expect(mocks.changeContext.mock.calls).toHaveLength(0); // from last time
    expect(mocks.changeData.mock.calls).toHaveLength(0); // from last time
  });

  it('change model prop - few times, while first change not done yet', async () => {
    // init and verify
    await init();

    const model2 = { id: 'test2', fields: { b: { path: 'b' } } };
    const model3 = { id: 'test2', fields: { b: { path: 'c' } } };

    // change model prop triggers destroy & init actions
    await act(async () => {
      props.model = model2;
      await root.update(<Form {...props}>{children}</Form>);

      props.model = model3;
      await root.update(<Form {...props}>{children}</Form>); 
    });
    expect(mocks.destroy.mock.calls).toHaveLength(2);
    expect(mocks.init.mock.calls).toHaveLength(3);
    expect(mocks.init.mock.calls[1]).toEqual([model2, props.resources, props.settings, expect.any(Function)]);
    expect(mocks.init.mock.calls[2]).toEqual([model3, props.resources, props.settings, expect.any(Function)]);
    expect(mocks.changeContext.mock.calls).toHaveLength(0);
    expect(mocks.changeData.mock.calls).toHaveLength(0);
    expect(mocksInvalid.init.mock.calls).toHaveLength(0);
  });

  it('change data prop - triggers changeData action', async () => {
    // init and verify
    await init();

    // change data prop - triggers changeData action
    props.data = { a: 1 };
    await act(async () => { await root.update(<Form {...props}>{children}</Form>); });
    expect(mocks.changeData).toHaveBeenCalledWith(props.data);
    expect(mocks.init.mock.calls).toHaveLength(1); // from last time
    expect(mocks.changeContext.mock.calls).toHaveLength(0); // from last time
    expect(mocks.destroy.mock.calls).toHaveLength(0); // from last time
  });

  it('change context prop - triggers changeContext action', async () => {
    // init and verify
    await init();

    // change context prop - triggers changeContext action
    props.context = { a: 1 };
    await act(async () => { await root.update(<Form {...props}>{children}</Form>); });
    expect(mocks.changeContext).toHaveBeenCalledWith(props.context);
    expect(mocks.init.mock.calls).toHaveLength(1); // from last time
    expect(mocks.changeData.mock.calls).toHaveLength(0); // from last time
    expect(mocks.destroy.mock.calls).toHaveLength(0); // from last time
  });

  it('unmount - triggers destroy action and stops receiving UI updates to the public form', async () => {
    // init and verify
    await init();

    // unmount
    await act(async () => { await root.unmount(); });

    // verify only destroy action called
    expect(mocks.destroy.mock.calls).toHaveLength(1);
    expect(mocks.init.mock.calls).toHaveLength(1); // from last time
    expect(mocks.changeData.mock.calls).toHaveLength(0); // from last time
    expect(mocks.changeContext.mock.calls).toHaveLength(0); // from last time

    // verify stops getting ui updates (the internal form triggers another ui updates such as DESTROY - END_PROCESSING
    // after the component un-mounted - and causing set state to unmounted component - which throws error)
    const updatePublicForm = mocks.init.mock.calls[0][3];
    const form = { model: undefined, settings: props.settings, resources: props.resources };
    const orgError = global.console.error;
    global.console.error = jest.fn();

    act(() => { updatePublicForm(form); });
    expect(global.console.error).not.toHaveBeenCalled();

    global.console.error = orgError;
  });

  it('unmount - after bad model init - not triggers destroy action and stops receiving UI updates to the public form',
    async () => {
    // init and verify
      await init();

      // change model prop to invalid model - triggers destroy & init actions
      JForm.default.prototype.init = mocksInvalid.init; // init finish not ok
      props.model = {};
      await act(async () => { await root.update(<Form {...props}>{children}</Form>); });
      expect(mocks.destroy.mock.calls).toHaveLength(1);
      expect(mocksInvalid.init.mock.calls).toHaveLength(1);
      expect(mocksInvalid.init.mock.calls[0]).toEqual([props.model, props.resources, props.settings, expect.any(Function)]);
      expect(mocks.init.mock.calls).toHaveLength(1); // from last time
      expect(mocks.changeContext.mock.calls).toHaveLength(0); // from last time
      expect(mocks.changeData.mock.calls).toHaveLength(0); // from last time

      // unmount
      await act(async () => { await root.unmount(); });

      // verify destroy action not called
      expect(mocks.destroy.mock.calls).toHaveLength(1); // from last time
      expect(mocks.init.mock.calls).toHaveLength(1); // from last time
      expect(mocks.changeData.mock.calls).toHaveLength(0); // from last time
      expect(mocks.changeContext.mock.calls).toHaveLength(0); // from last time

      // verify stops getting ui updates (the internal form triggers another ui updates such as DESTROY - END_PROCESSING
      // after the component un-mounted - and causing set state to unmounted component - which throws error)
      const updatePublicForm = mocks.init.mock.calls[0][3];
      const form = { model: undefined, settings: props.settings, resources: props.resources };
      const orgError = global.console.error;
      global.console.error = jest.fn();

      act(() => { updatePublicForm(form); });
      expect(global.console.error).not.toHaveBeenCalled();

      global.console.error = orgError;
    });
});
