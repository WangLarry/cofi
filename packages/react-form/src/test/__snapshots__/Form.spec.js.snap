// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form init - triggers init action and update public ui form 1`] = `
<div>
  Mock Child
</div>
`;
